### PR TITLE
Convert NFS and Cassandra CRDs to v1 and auto-generate the CRDs

### PR DESCRIPTION
**Pull request by travisn**
**Description of your changes:**
The CRDs v1 requires the full schema for all settings. The deprecation coming for CRDs v1beta1 coming in K8s 1.22 we need to update all the providers to the v1 schema. Now we consistently generate the CRDs from the types for all providers.

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: build/crds/build-crds.sh [hunk 1]
```diff
@@ -39,10 +40,20 @@ copy_ob_obc_crds() {
...
+
+  echo "Generating cassandra crds"
+  "$CONTROLLER_GEN_BIN_PATH" "$CRD_OPTIONS" paths="./pkg/apis/cassandra.rook.io/v1alpha1" output:crd:artifacts:config="$CASSANDRA_CRDS_DIR"
+  # Format with yq for consistent whitespace
+  $YQ_BIN_PATH read $CASSANDRA_CRDS_DIR/cassandra.rook.io_clusters.yaml > $CASSANDRA_CRDS_DIR/crds.yaml
+  rm -f $CASSANDRA_CRDS_DIR/cassandra.rook.io_clusters.yaml
+
+  echo "Generating nfs crds"
+  "$CONTROLLER_GEN_BIN_PATH" "$CRD_OPTIONS" paths="./pkg/apis/nfs.rook.io/v1alpha1" output:crd:artifacts:config="$NFS_CRDS_DIR"
+  mv $NFS_CRDS_DIR/nfs.rook.io_nfsservers.yaml  $NFS_CRDS_DIR/crds.yaml
```
Comment (leseb) [inline build/crds/build-crds.sh]: There is an extra space between the 2 arguments.

File: build/crds/build-crds.sh [hunk 2]
```diff
@@ -39,10 +40,20 @@ copy_ob_obc_crds() {
...
   $YQ_BIN_PATH w -i cluster/olm/ceph/deploy/crds/ceph.rook.io_cephclusters.yaml spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.storage.properties.storageClassDeviceSets.items.properties.volumeClaimTemplates.items.properties.metadata.properties.annotations.x-kubernetes-preserve-unknown-fields true
+
+  echo "Generating cassandra crds"
+  "$CONTROLLER_GEN_BIN_PATH" "$CRD_OPTIONS" paths="./pkg/apis/cassandra.rook.io/v1alpha1" output:crd:artifacts:config="$CASSANDRA_CRDS_DIR"
+  # Format with yq for consistent whitespace
+  $YQ_BIN_PATH read $CASSANDRA_CRDS_DIR/cassandra.rook.io_clusters.yaml > $CASSANDRA_CRDS_DIR/crds.yaml
+  rm -f $CASSANDRA_CRDS_DIR/cassandra.rook.io_clusters.yaml
+
+  echo "Generating nfs crds"
+  "$CONTROLLER_GEN_BIN_PATH" "$CRD_OPTIONS" paths="./pkg/apis/nfs.rook.io/v1alpha1" output:crd:artifacts:config="$NFS_CRDS_DIR"
```
Comment (leseb) [inline build/crds/build-crds.sh:56]: No `yq` read?

File: pkg/apis/cassandra.rook.io/v1alpha1/types.go [hunk 3]
```diff
@@ -52,8 +52,10 @@ const (
 type Cluster struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata"`
-	Spec              ClusterSpec   `json:"spec"`
-	Status            ClusterStatus `json:"status"`
+	Spec              ClusterSpec `json:"spec"`
+	// +optional
+	// +nullable
+	Status ClusterStatus `json:"status"`
```
Comment (leseb) [inline pkg/apis/cassandra.rook.io/v1alpha1/types.go:58]: You need `,omitempty"` too

File: pkg/apis/cassandra.rook.io/v1alpha1/types.go [hunk 4]
```diff
@@ -67,16 +69,25 @@ type ClusterList struct {
 // ClusterSpec is the desired state for a Cassandra Cluster.
 type ClusterSpec struct {
 	// The annotations-related configuration to add/set on each Pod related object.
+	// +optional
+	// +nullable
 	Annotations rook.Annotations `json:"annotations"`
```
Comment (leseb) [inline pkg/apis/cassandra.rook.io/v1alpha1/types.go:74]: You need `,omitempty"` too

File: pkg/apis/cassandra.rook.io/v1alpha1/types.go [hunk 5]
```diff
@@ -67,16 +69,25 @@ type ClusterList struct {
...
+	// +optional
+	// +nullable
 	Repository *string `json:"repository,omitempty"`
 	// Mode selects an operating mode.
+	// +optional
 	Mode ClusterMode `json:"mode,omitempty"`
 	// Datacenter that will make up this cluster.
+	// +optional
+	// +nullable
 	Datacenter DatacenterSpec `json:"datacenter"`
```
Comment (leseb) [inline pkg/apis/cassandra.rook.io/v1alpha1/types.go:87]: You need `,omitempty"` too

File: pkg/apis/cassandra.rook.io/v1alpha1/types.go [hunk 6]
```diff
@@ -102,16 +113,26 @@ type RackSpec struct {
...
 	// User-provided ConfigMap for jmx prometheus exporter
+	// +optional
+	// +nullable
 	JMXExporterConfigMapName *string `json:"jmxExporterConfigMapName,omitempty"`
 	// Storage describes the underlying storage that Cassandra will consume.
 	Storage StorageScopeSpec `json:"storage"`
 	// The annotations-related configuration to add/set on each Pod related object.
+	// +optional
+	// +nullable
 	Annotations map[string]string `json:"annotations"`
```
Comment (leseb) [inline pkg/apis/cassandra.rook.io/v1alpha1/types.go:128]: You need `,omitempty"` too

File: pkg/apis/cassandra.rook.io/v1alpha1/types.go [hunk 7]
```diff
@@ -102,16 +113,26 @@ type RackSpec struct {
...
 	// Placement describes restrictions for the nodes Cassandra is scheduled on.
+	// +optional
+	// +nullable
 	Placement *Placement `json:"placement,omitempty"`
 	// Resources the Cassandra Pods will use.
+	// +optional
+	// +nullable
 	Resources corev1.ResourceRequirements `json:"resources"`
```
Comment (leseb) [inline pkg/apis/cassandra.rook.io/v1alpha1/types.go:136]: You need `,omitempty"` too

File: pkg/apis/cassandra.rook.io/v1alpha1/types.go [hunk 8]
```diff
@@ -120,6 +141,7 @@ type ImageSpec struct {
 	// Version of the image.
 	Version string `json:"version"`
 	// Repository to pull the image from.
+	// +optional
 	Repository string `json:"repository"`
```
Comment (leseb) [inline pkg/apis/cassandra.rook.io/v1alpha1/types.go:145]: You need `,omitempty"` too

**Review by leseb — CHANGES_REQUESTED** [review]

File: build/crds/build-crds.sh (hunk 2 above)
Comment (travisn) [reply build/crds/build-crds.sh:56]: There is a `yq read` on the next line, is that what you mean?

Comment (leseb) [reply build/crds/build-crds.sh:56]: Oh wow I must have missed it. Never mind!

**Review by leseb — APPROVED** [review]

**Pull request merged**